pymatgen/io/input.py

```diff
@@ -0,0 +1,172 @@
+                    try:
+                        zip.write(file)
+                        os.remove(file)
+                    except FileNotFoundError:
+                        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_directory(cls, directory: Union[str, Path]):
+        """
+        Construct an InputSet from a directory of one or more files.
+
+        Args:
+            directory: Directory to read input files from
+        """
+        raise NotImplementedError(f"from_directory has not been implemented in {cls}")
+
+    def validate(self) -> bool:
+        """
+        A place to implement basic checks to verify the validity of an
+        input set. Can be as simple or as complex as desired.
+
+        Will raise a NotImplementedError unless overloaded by the inheriting class.
+        """
+        raise NotImplementedError(f".validate() has not been implemented in {self.__class__}")
+
+
+class InputSetGenerator(MSONable):
+    """
+    InputSetGenerator classes serve as generators for Input objects. They contain
+    settings or sets of instructions for how to create Input from a set of
+    coordinates or a previous calculation directory.
+    """
+
+    @abc.abstractmethod
+    def get_input_set(self) -> InputSet:
+        """
+        Generate an InputSet object. Typically the first argument to this method
+        will be a Structure or other form of atomic coordinates.
+        """
+        pass
+
+
+class TemplateInputSet(InputSet):
+    """
+    Concrete implementation of InputSet that is based on a single template input
+    file with variables.
+
+    This class is provided as a low-barrier way to support new codes and to provide
+    an intuitive way for users to transition from manual scripts to pymatgen I/O
+    classes.
+    """
+
+    def __init__(self, template: Union[str, Path], variables: Optional[Dict] = None):
+        """
+        Args:
+            template: the input file template containing variable strings to be
+                replaced.
+            variables: dict of variables to replace in the template. Keys are the
+                text to replaced with the values, e.g. {"TEMPERATURE": 298} will
+                replace the text $TEMPERATURE in the template. See Python's
+                Template.safe_substitute() method documentation for more details.
+        """
+        self.template = template
+        self.variables = variables if variables else {}
+
+        # load the template
+        with zopen(self.template, "r") as f:
+            template_str = f.read()
+
+        # replace all variables
+        self.data = Template(template_str).safe_substitute(**self.variables)
+
+    def _generate_input_data(self, filename: str = "input.txt"):
+        """
+        Args:
+            filename: name of the file to be written
+        """
+        return {filename: self.data}
+
+    @classmethod
+    def from_directory(cls, directory: Union[str, Path]):
```

Comment:
I think this should be easily to implement via comparisons to determine the variables?

Reply:
If it is a template, 

```
key: $value
```

and actual input says

```
key: hello
```

Then it follows that $value = "hello"?